asset_dashboard/templates/asset_dashboard/partials/forms/add_edit_phase_form.html

```diff
@@ -129,6 +154,239 @@ <h3>Phase Assets</h3>
+            error.setAttribute("aria-hidden","true")
+          })
+
+          let data = {
+            'year': funding[0].value,
+            'budget': funding[1].value,
+            'actual_cost': funding[3].value,
+            'source_type': funding[5].value,
+            'funding_secured': funding[6].value,
+            'phase': props.phase_id,
+            'id': funding.id ? funding.id : null,
+          }
+
+          // Get all fetches into an array, so we can wait to check
+          // if they were successful after they're all done
+          fetches.push(
+            fetch(`/projects/phases/fundingstream/`, {
+              method: 'POST',
+              body: JSON.stringify(data),
+              headers: {
+                'X-CSRFToken': getCookie('csrftoken'),
+                'Accept': 'application/json',
+                'Content-Type': 'application/json',
+              },
+              mode: 'same-origin' 
+            }).then((response)=> {
+              if (response.status != 201) {
+                backendValid = false
+              }
+              return response.json()
+            }).then(data => {
+              if (!backendValid) {
+                showBackendErrors(data, ...errors)
+              }
+            })
+          )
+        }
+
+        // Wait until all fetches run, then either show serverError or reload page
+        Promise.all(fetches).then(function() {
+          if (!backendValid) {
+            showError(serverError)
+          } else {
+            hideError(serverError)
+            window.removeEventListener('beforeunload', redirectDialog)
+            location.reload()
+          }
+        })    
+      }
+    }
+
+    const getCookie = (name) => {
+        const match = document.cookie.split(';').map(el => { let [key,value] = el.split('='); return { [key.trim()]: value }})
+        const filteredMatch = match.filter(e => Object.keys(e)[0] === name)
+
+        let matchLength = filteredMatch.length
+
+        return filteredMatch[matchLength - 1][name]
+    }
+
+    window.addEventListener('beforeunload', redirectDialog)
+
+    function redirectDialog(e){
+      e.preventDefault();
+      let hasUnsavedForms = false
+      let newFormsContainers = document.getElementsByClassName('new-tr')
+
+      for(i = 0; i < newFormsContainers.length; i++ ){
+        let yearInput = newFormsContainers[i].querySelector('#id_year').value
+        let budgetInput = newFormsContainers[i].querySelector('#id_budget_0').value
+        let costInput = newFormsContainers[i].querySelector('#id_actual_cost_0').value
+        
+        if(yearInput || budgetInput > 0 || costInput > 0) {
+          // Ask user if they want to leave page, if there are unsaved funding inputs
+          e.returnValue = ""
+        }
+          
+      }
+    }
+
+    function showBackendErrors(data, yearError, budgetError, costError, sourceError, securedError) {
+      let keys = Object.keys(data)
+      keys.forEach(key => {
```

Comment:
you can just do `Object.keys(data).forEach` without creating the keys variable, but that's a stylistic thing that is up to you.